### PR TITLE
Update dependencies for OpenSSL 1.1.0 compatibility

### PR DESCRIPTION
**Pull request by cuviper**
The primary targets here are openssl and openssl-sys crates 0.9,
bringing support for OpenSSL 1.1.0.  This requires updating the curl
and git2 related dependencies as well.

A small change is required in cargo itself for the new Hasher API.
Results from the hasher are simply unwrapped for now, matching the
Windows behavior that already panics on error.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I've been passively trying to do this but the git2 behavior changed linking to libcurl enough that it's been difficult to find time to fix all the tests. It looks like the Travis builds may also be failing?

**Comment by cuviper** [comment]
Sorry, simply missing the new `OPENSSL_DIR` was the main problem for Travis.  Linux is all green now, but OSX still failed, and I don't have the means to debug that myself.

What are the git2/libcurl problems you're referring to?

Anyway, I've opened the PR to maintainer pushes if you'd like to add anything, however passively. :)

File: .travis.yml [hunk 1]
```diff
@@ -25,6 +25,7 @@ matrix:
            MACOSX_DEPLOYMENT_TARGET=10.7
       os: osx
       before_install:
+        - export OPENSSL_DIR=`brew --prefix openssl`
         - export OPENSSL_INCLUDE_DIR=`brew --prefix openssl`/include
```
Comment (alexcrichton) [inline .travis.yml]: I think these env vars can be removed now

File: Cargo.toml [hunk 2]
```diff
@@ -26,12 +26,12 @@ env_logger = "0.3"
 filetime = "0.1"
 flate2 = "0.2"
 fs2 = "0.2"
-git2 = "0.4"
-git2-curl = "0.5"
+git2 = "0.5"
```
Comment (alexcrichton) [inline Cargo.toml]: Ah FWIW I had to end up yanking the 0.5 release that had support for openssl 1.1, so this'll need to be updated to 0.6 now

File: Makefile.in [hunk 3]
```diff
@@ -251,10 +251,12 @@ target/openssl/$(1).stamp: target/openssl/openssl-$$(OPENSSL_VERS).tar.gz \
 
 # variables read by various build scripts to find openssl
 cargo-$(1): export OPENSSL_STATIC := 1
+cargo-$(1): export OPENSSL_DIR := $$(OPENSSL_INSTALL_$(1))
 cargo-$(1): export OPENSSL_ROOT_DIR := $$(OPENSSL_INSTALL_$(1))
 cargo-$(1): export OPENSSL_LIB_DIR := $$(OPENSSL_INSTALL_$(1))/lib
 cargo-$(1): export OPENSSL_INCLUDE_DIR := $$(OPENSSL_INSTALL_$(1))/include
```
Comment (alexcrichton) [inline Makefile.in]: lib/include here I believe can be removed

**Comment by alexcrichton** [comment]
Ok, pushed a number of updates and fixes, the remaining errors seem easy to deal with, just need to figure them out.

**Comment by cuviper** [comment]
The AppVeyor failure couldn't remove a build directory -- possibly spurious?

The OSX failure got a signal 4, SIGILL? -- don't know what to think about that.

**Comment by alexcrichton** [comment]
Ah yeah AppVeyor looks spurious, didn't realize that OSX was signal 4... Odd! I'll poke around locally.

**Comment by alexcrichton** [comment]
Looks like a legit segfault in libgit2, going to try to fix

**Comment by alexcrichton** [comment]
I've [submitted a PR](https://github.com/libgit2/libgit2/pull/3993) to fix the fault in libgit2

**Comment by alexcrichton** [comment]
Ok, fix merged upstream in libgit2, and libgit2-sys is updated. Pushed and let's see what CI think.

**Comment by alexcrichton** [comment]
@bors: r+

Alright, let's see if it lands!

**Comment by sfackler** [comment]
FYI, you'll need to upgrade to rust-openssl 0.9.1 to fix a bug that broke EOF detection in OpenSSL 1.1.0c if you're using `SslStream`.

**Comment by alexcrichton** [comment]
@bors: retry

@sfackler oh we're not actually using that at all yet in Cargo, just the openssl-sys crate to link to OpenSSL itself. Cargo only very lightly uses OpenSSL to do a bit of hashing, but other than that the `openssl` crate itself isn't used.

**Comment by cuviper** [comment]
Not even in the transitive dependencies?  The version here is chosen for everyone, after all...

**Comment by alexcrichton** [comment]
Yep, as the author of most of them, I can tell you the `openssl` crate basically isn't used.

**Comment by cuviper** [comment]
Heh, fair enough.

**Comment by alexcrichton** [comment]
Looks like pkg-config needs to be installed on the bots, I'll work on getting that done.

**Comment by alexcrichton** [comment]
@bors: r+

Probably gonna fail, but I'm curious to see the new error...

**Comment by cuviper** [comment]
Is it normal for cargo testing to be so fickle?

**Comment by cuviper** [comment]
@alexcrichton Thanks for hammering that through!

**Pull request merged**